src/features/jserrors/instrument/index.js

```diff
@@ -5,64 +5,59 @@
-import { wrapRaf, wrapTimer, wrapEvents, wrapXhr } from '../../../common/wrap'
-import './debug'
 import { InstrumentBase } from '../../utils/instrument-base'
-import { FEATURE_NAME, NR_ERR_PROP } from '../constants'
+import { FEATURE_NAME } from '../constants'
 import { FEATURE_NAMES } from '../../../loaders/features/features'
 import { globalScope } from '../../../common/constants/runtime'
 import { eventListenerOpts } from '../../../common/event-listener/event-listener-opts'
-import { getRuntime } from '../../../common/config/config'
 import { stringify } from '../../../common/util/stringify'
+import { UncaughtError } from './uncaught-error'
 
 export class Instrument extends InstrumentBase {
   static featureName = FEATURE_NAME
+
+  #seenErrors = new Set()
+
   constructor (agentIdentifier, aggregator, auto = true) {
     super(agentIdentifier, aggregator, FEATURE_NAME, auto)
-    // skipNext counter to keep track of uncaught
-    // errors that will be the same as caught errors.
-    this.skipNext = 0
+
     try {
       // this try-catch can be removed when IE11 is completely unsupported & gone
       this.removeOnAbort = new AbortController()
     } catch (e) {}
 
-    const thisInstrument = this
-    thisInstrument.ee.on('fn-start', function (args, obj, methodName) {
-      if (thisInstrument.abortHandler) thisInstrument.skipNext += 1
-    })
-    thisInstrument.ee.on('fn-err', function (args, obj, err) {
-      if (thisInstrument.abortHandler && !err[NR_ERR_PROP]) {
-        getOrSet(err, NR_ERR_PROP, function getVal () {
-          return true
-        })
-        this.thrown = true
-        handle('err', [err, now()], undefined, FEATURE_NAMES.jserrors, thisInstrument.ee)
-      }
-    })
-    thisInstrument.ee.on('fn-end', function () {
-      if (!thisInstrument.abortHandler) return
-      if (!this.thrown && thisInstrument.skipNext > 0) thisInstrument.skipNext -= 1
+    // Capture function errors early in case the spa feature is loaded
+    this.ee.on('fn-err', (args, obj, error) => {
+      if (!this.abortHandler || this.#seenErrors.has(error)) return
+      this.#seenErrors.add(error)
+
+      handle('err', [this.#castError(error), now()], undefined, FEATURE_NAMES.jserrors, this.ee)
     })
-    thisInstrument.ee.on('internal-error', function (e) {
-      handle('ierr', [e, now(), true], undefined, FEATURE_NAMES.jserrors, thisInstrument.ee)
+
+    this.ee.on('internal-error', (error) => {
+      if (!this.abortHandler) return
+      handle('ierr', [this.#castError(error), now(), true], undefined, FEATURE_NAMES.jserrors, this.ee)
     })
 
-    // Replace global error handler with our own.
-    this.origOnerror = globalScope.onerror
-    globalScope.onerror = this.onerrorHandler.bind(this)
+    globalScope.addEventListener('unhandledrejection', (promiseRejectionEvent) => {
+      if (!this.abortHandler) return
+      if (this.#seenErrors.has(promiseRejectionEvent.reason)) {
```

Comment:
Does I didn't think fn-err fired in the promise wrapper.  Is this validated?

Reply:
if yes and it ends up getting sent as an error in the fn-err handler, its not getting marked as a promise exception

Reply:
Verified the promise wrapper does not use the `fn-err` event so unhandled promise rejections will get sent as such.
<img width="685" alt="image" src="https://github.com/newrelic/newrelic-browser-agent/assets/3984979/7a58b3ce-d1ad-4157-8746-74f1673cf3d9">

Reply:
I guess technically the `seenErrors` bit here is not necessary. I will remove it.